File: 1443. Minimum Time to Collect All Apples in a Tree/parent.cpp
class Solution {
public:
    int minTime(int n, vector<vector<int>>& edges, vector<bool>& hasApple){
        if (find(hasApple.begin(), hasApple.end(), true) == hasApple.end())
            return 0;

        vector<int> parents(n, -1);

        for (int i = 0; i < edges.size(); i++){
            int parent = edges[i][0];
            int child = edges[i][1];

            if (parents[child] == -1)
                parents[child] = parent;
            else
                parents[parent] = child;

        }

        for (int i = n - 1; i > 0; i--) {
            if ((hasApple[i] == true) && (parents[i] != -1))
                hasApple[parents[i]] = 1;
        }

        int res = 0;

        for (int i = 1; i < n; i++)
            res += hasApple[i];

        return res * 2;
    }
};

File: 1443. Minimum Time to Collect All Apples in a Tree/parent.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        if True not in hasApple:
            return 0

        parents = [-1] * n

        for edge in edges:
            parent, child = edge

            if parents[child] == -1:
                parents[child] = parent
            else:
                parents[parent] = child

        for index in range(n - 1, 0, -1):
            if hasApple[index] and parents[index] != -1:
                hasApple[parents[index]] = 1

        return sum(hasApple[1:]) * 2
